In [1]:
# 노트북 안에서 설치가 필요한 경우에만 아래 명령어의 주석을 해제하고 실행합니다.
# 단, 현재 선택된 Jupyter 커널이 uv 가상환경인지 먼저 확인해야 합니다.
#!pip install pandas numpy matplotlib seaborn folium ipywidgets scikit-learn

In [2]:
%matplotlib inline

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# seaborn: heatmap, barplot처럼 보기 좋은 그래프를 그릴 때 사용합니다.
import seaborn as sns

# folium: 위도/경도 데이터를 지도 위에 표시할 때 사용합니다.
import folium

# HeatMap: folium 지도 위에 데이터 밀집 정도를 색으로 표현하는 도구입니다.
from folium.plugins import HeatMap

In [3]:
plt.style.use('ggplot')

In [4]:
# 한글 폴트 설정 (Windows 기준). 폰트가 없거나 다른 OS에서 실행해도 오류가 나지 않도록 예외 처리합니다.
try:
    plt.rcParams['font.family'] = 'Malgun Gothic'
    plt.rcParams['axes.unicode_minus'] = False
except Exception as e:
    print('한글 폰트 설정을 건너뜁니다:', e)

In [5]:
df = pd.read_csv('jeju_bus.csv')
df

,id,date,route_id,vh_id,route_nm,now_latitude,now_longitude,now_station,now_arrive_time,distance,next_station,next_latitude,next_longitude,next_arrive_time
0,0,2019-10-15,405136001,7997025,360-1,33.456267,126.551750,제주대학교입구,06시,266.0,제대마을,33.457724,126.554014,24
1,1,2019-10-15,405136001,7997025,360-1,33.457724,126.554014,제대마을,06시,333.0,제대아파트,33.458783,126.557353,36
2,2,2019-10-15,405136001,7997025,360-1,33.458783,126.557353,제대아파트,06시,415.0,제주대학교,33.459893,126.561624,40
3,3,2019-10-15,405136001,7997025,360-1,33.479705,126.543811,남국원(아라방면),06시,578.0,제주여자중고등학교(아라방면),33.484860,126.542928,42
4,4,2019-10-15,405136001,7997025,360-1,33.485662,126.494923,도호동,07시,374.0,은남동,33.485822,126.490897,64
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210452,210452,2019-10-28,405328102,7983486,281-2,33.255783,126.577450,비석거리,21시,528.0,삼아아파트,33.251896,126.574417,96
210453,210453,2019-10-28,405328102,7983486,281-2,33.248595,126.568527,동문로터리,21시,280.0,매일올레시장 7번입구,33.249753,126.565959,50
210454,210454,2019-10-28,405328102,7983486,281-2,33.251891,126.560303,서귀포시 구 버스터미널,21시,114.0,아랑조을거리 입구,33.251084,126.559551,16
210455,210455,2019-10-28,405328102,7983486,281-2,33.251084,126.559551,아랑조을거리 입구,21시,223.0,평생학습관,33.249504,126.558068,38


In [6]:
df

,id,date,route_id,vh_id,route_nm,now_latitude,now_longitude,now_station,now_arrive_time,distance,next_station,next_latitude,next_longitude,next_arrive_time
0,0,2019-10-15,405136001,7997025,360-1,33.456267,126.551750,제주대학교입구,06시,266.0,제대마을,33.457724,126.554014,24
1,1,2019-10-15,405136001,7997025,360-1,33.457724,126.554014,제대마을,06시,333.0,제대아파트,33.458783,126.557353,36
2,2,2019-10-15,405136001,7997025,360-1,33.458783,126.557353,제대아파트,06시,415.0,제주대학교,33.459893,126.561624,40
3,3,2019-10-15,405136001,7997025,360-1,33.479705,126.543811,남국원(아라방면),06시,578.0,제주여자중고등학교(아라방면),33.484860,126.542928,42
4,4,2019-10-15,405136001,7997025,360-1,33.485662,126.494923,도호동,07시,374.0,은남동,33.485822,126.490897,64
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210452,210452,2019-10-28,405328102,7983486,281-2,33.255783,126.577450,비석거리,21시,528.0,삼아아파트,33.251896,126.574417,96
210453,210453,2019-10-28,405328102,7983486,281-2,33.248595,126.568527,동문로터리,21시,280.0,매일올레시장 7번입구,33.249753,126.565959,50
210454,210454,2019-10-28,405328102,7983486,281-2,33.251891,126.560303,서귀포시 구 버스터미널,21시,114.0,아랑조을거리 입구,33.251084,126.559551,16
210455,210455,2019-10-28,405328102,7983486,281-2,33.251084,126.559551,아랑조을거리 입구,21시,223.0,평생학습관,33.249504,126.558068,38


In [7]:
df['now_arrive_time'].str.replace('시', '', regex=False)

0         06
1         06
2         06
3         06
4         07
          ..
210452    21
210453    21
210454    21
210455    21
210456    21
Name: now_arrive_time, Length: 210457, dtype: str

In [8]:
# regex: 정규식 여부 확인
df['now_hour'] = df['now_arrive_time'].str.replace('시', '', regex=False).astype(int)

In [9]:
df[ ['now_arrive_time', 'now_hour'] ]

,now_arrive_time,now_hour
0,06시,6
1,06시,6
2,06시,6
3,06시,6
4,07시,7
...,...,...
210452,21시,21
210453,21시,21
210454,21시,21
210455,21시,21


In [10]:
df['date'] = pd.to_datetime( df['date'] )

In [11]:
df['dayofweek'] = df['date'].dt.dayofweek

In [12]:
df

,id,date,route_id,vh_id,route_nm,now_latitude,now_longitude,now_station,now_arrive_time,distance,next_station,next_latitude,next_longitude,next_arrive_time,now_hour,dayofweek
0,0,2019-10-15,405136001,7997025,360-1,33.456267,126.551750,제주대학교입구,06시,266.0,제대마을,33.457724,126.554014,24,6,1
1,1,2019-10-15,405136001,7997025,360-1,33.457724,126.554014,제대마을,06시,333.0,제대아파트,33.458783,126.557353,36,6,1
2,2,2019-10-15,405136001,7997025,360-1,33.458783,126.557353,제대아파트,06시,415.0,제주대학교,33.459893,126.561624,40,6,1
3,3,2019-10-15,405136001,7997025,360-1,33.479705,126.543811,남국원(아라방면),06시,578.0,제주여자중고등학교(아라방면),33.484860,126.542928,42,6,1
4,4,2019-10-15,405136001,7997025,360-1,33.485662,126.494923,도호동,07시,374.0,은남동,33.485822,126.490897,64,7,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210452,210452,2019-10-28,405328102,7983486,281-2,33.255783,126.577450,비석거리,21시,528.0,삼아아파트,33.251896,126.574417,96,21,0
210453,210453,2019-10-28,405328102,7983486,281-2,33.248595,126.568527,동문로터리,21시,280.0,매일올레시장 7번입구,33.249753,126.565959,50,21,0
210454,210454,2019-10-28,405328102,7983486,281-2,33.251891,126.560303,서귀포시 구 버스터미널,21시,114.0,아랑조을거리 입구,33.251084,126.559551,16,21,0
210455,210455,2019-10-28,405328102,7983486,281-2,33.251084,126.559551,아랑조을거리 입구,21시,223.0,평생학습관,33.249504,126.558068,38,21,0


In [13]:
plt.style.use('ggplot')

plt.figure(figsize=(15, 8) )
plt.hist( df['next_arrive_time'], bins = 100 )

plt.figure(figsize = (20, 8))
# vert: 수직 수평 여부
plt.boxplot( df['next_arrive_time'], vert=False )

In [16]:
Q1 = df['next_arrive_time'].quantile(0.25)

Q1

np.float64(44.0)

In [17]:
Q2 = df['next_arrive_time'].quantile(0.5)
Q2

np.float64(66.0)

In [18]:
Q3 = df['next_arrive_time'].quantile(0.75)
Q3

np.float64(102.0)

In [19]:
IQR = Q3 - Q1
IQR

np.float64(58.0)

In [20]:
Q1 - 1.5 * IQR

np.float64(-43.0)

In [21]:
Q3 + 1.5 * IQR

np.float64(189.0)

In [22]:
# 이상치 확인
q90 = df['next_arrive_time'].quantile(0.90)

q90

np.float64(154.0)

In [23]:
df_under_q90 = df[ df['next_arrive_time'] <= q90 ]

In [24]:
df_under_q90

,id,date,route_id,vh_id,route_nm,now_latitude,now_longitude,now_station,now_arrive_time,distance,next_station,next_latitude,next_longitude,next_arrive_time,now_hour,dayofweek
0,0,2019-10-15,405136001,7997025,360-1,33.456267,126.551750,제주대학교입구,06시,266.0,제대마을,33.457724,126.554014,24,6,1
1,1,2019-10-15,405136001,7997025,360-1,33.457724,126.554014,제대마을,06시,333.0,제대아파트,33.458783,126.557353,36,6,1
2,2,2019-10-15,405136001,7997025,360-1,33.458783,126.557353,제대아파트,06시,415.0,제주대학교,33.459893,126.561624,40,6,1
3,3,2019-10-15,405136001,7997025,360-1,33.479705,126.543811,남국원(아라방면),06시,578.0,제주여자중고등학교(아라방면),33.484860,126.542928,42,6,1
4,4,2019-10-15,405136001,7997025,360-1,33.485662,126.494923,도호동,07시,374.0,은남동,33.485822,126.490897,64,7,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210452,210452,2019-10-28,405328102,7983486,281-2,33.255783,126.577450,비석거리,21시,528.0,삼아아파트,33.251896,126.574417,96,21,0
210453,210453,2019-10-28,405328102,7983486,281-2,33.248595,126.568527,동문로터리,21시,280.0,매일올레시장 7번입구,33.249753,126.565959,50,21,0
210454,210454,2019-10-28,405328102,7983486,281-2,33.251891,126.560303,서귀포시 구 버스터미널,21시,114.0,아랑조을거리 입구,33.251084,126.559551,16,21,0
210455,210455,2019-10-28,405328102,7983486,281-2,33.251084,126.559551,아랑조을거리 입구,21시,223.0,평생학습관,33.249504,126.558068,38,21,0


plt.figure(figsize=(20,8))

# s: 원 크기
# alpha: 투명도
plt.scatter(
    x = df_under_q90['distance'],
    y = df_under_q90['next_arrive_time'],
    s = 5,
    alpha = 0.2
    )

In [26]:
(df['next_arrive_time'] > 300).sum()

np.int64(3070)

In [27]:
df[ ['distance', 'next_arrive_time', 'now_hour'] ]

,distance,next_arrive_time,now_hour
0,266.0,24,6
1,333.0,36,6
2,415.0,40,6
3,578.0,42,6
4,374.0,64,7
...,...,...,...
210452,528.0,96,21
210453,280.0,50,21
210454,114.0,16,21
210455,223.0,38,21


corr_df = pd.DataFrame({
              'kor' : [70, 88, 90, 95, 100],
              'eng' : [65, 70, 78, 85, 96],
              'gym' : [98, 87, 72, 78, 80],
           'height' : [1.8, 1.6, 1.7, 1.45, 1.72],
           'weight' : [90, 80, 100, 85, 90]
})

corr_df.corr()

In [28]:
df[['now_hour', 'next_arrive_time']]

,now_hour,next_arrive_time
0,6,24
1,6,36
2,6,40
3,6,42
4,7,64
...,...,...
210452,21,96
210453,21,50
210454,21,16
210455,21,38


In [29]:
df [ df['now_hour'] == 5 ].sort_values(by = 'next_arrive_time', ascending = False ).head(30)

,id,date,route_id,vh_id,route_nm,now_latitude,now_longitude,now_station,now_arrive_time,distance,next_station,next_latitude,next_longitude,next_arrive_time,now_hour,dayofweek
66194,66194,2019-10-19,405136522,7997022,365-22,33.483527,126.475958,노형초등학교,05시,347.0,정존마을,33.481317,126.473300,2468,5,5
157456,157456,2019-10-25,405136522,7997056,365-22,33.483527,126.475958,노형초등학교,05시,347.0,정존마을,33.481317,126.473300,2456,5,4
127314,127314,2019-10-23,405136522,7997056,365-22,33.483527,126.475958,노형초등학교,05시,347.0,정존마을,33.481317,126.473300,2176,5,2
66462,66462,2019-10-19,405136522,7997056,365-22,33.483527,126.475958,노형초등학교,05시,347.0,정존마을,33.481317,126.473300,2048,5,5
78568,78568,2019-10-20,405136002,7997082,360-2,33.477560,126.477406,탐라도서관,05시,518.0,제주고등학교/중흥S클래스,33.476260,126.481417,1946,5,6
33053,33053,2019-10-17,405136002,7997082,360-2,33.477560,126.477406,탐라도서관,05시,518.0,제주고등학교/중흥S클래스,33.476260,126.481417,1940,5,3
107827,107827,2019-10-22,405136001,7997034,360-1,33.459893,126.561624,제주대학교,05시,415.0,제대아파트,33.458783,126.557353,1802,5,1
184493,184493,2019-10-27,405136012,7997041,360-12,33.458900,126.557300,제대아파트,05시,435.0,제주대학교,33.460114,126.561659,1616,5,6
1840,1840,2019-10-15,405136002,7997082,360-2,33.458900,126.557300,제대아파트,05시,435.0,제주대학교,33.460114,126.561659,1372,5,1
156049,156049,2019-10-25,405136521,7997001,365-21,33.458900,126.557300,제대아파트,05시,435.0,제주대학교,33.460114,126.561659,1332,5,4


In [30]:
df.groupby('now_hour')['next_arrive_time'].count()

now_hour
0         5
5       487
6      8409
7     13581
8     13838
9     13700
10    13011
11    12524
12    13433
13    13197
14    12592
15    12846
16    13476
17    13090
18    12987
19    13275
20    11852
21    11230
22     6470
23      454
Name: next_arrive_time, dtype: int64

In [31]:
df.groupby('now_hour')['next_arrive_time'].mean()

now_hour
0      56.000000
5     132.956879
6      85.986443
7      85.488697
8      87.666715
9      82.414015
10     82.623780
11     83.547988
12     83.550733
13     84.070471
14     84.036134
15     84.904250
16     86.759573
17     90.705271
18     94.958882
19     87.240979
20     83.124705
21     82.479252
22     75.355796
23     63.387665
Name: next_arrive_time, dtype: float64

In [32]:
df.groupby('now_hour')['next_arrive_time'].agg(
                      ['count', 'mean', 'median']
)

,count,mean,median
now_hour,,,
0,5,56.000000,36.0
5,487,132.956879,66.0
6,8409,85.986443,64.0
7,13581,85.488697,66.0
8,13838,87.666715,66.0
9,13700,82.414015,64.0
10,13011,82.623780,66.0
11,12524,83.547988,66.0
12,13433,83.550733,66.0


In [33]:
df['station_segment'] = df['now_station'] + ' → ' + df['next_station']

In [34]:
segment_mean_time = df.groupby('station_segment')['next_arrive_time'].mean()

In [35]:
segment_mean_time.sort_values(ascending = False).head(30)

station_segment
노형초등학교 → 정존마을                         2182.800000
부영아파트 → 제주고등학교/중흥S클래스                 2166.000000
서귀포여자고등학교 → 수모루                       2160.000000
노형오거리 → 노형초등학교                        1773.777778
사회복지법인 춘강 → 산천단 한국폴리텍대학               1478.000000
정존마을 → 제주한라대학교                        1253.000000
한국병원 → 제주버스터미널(가상정류소)                 1125.333333
제주고등학교/중흥S클래스 → 탐라도서관                  634.500000
탐라도서관 → 제주고등학교/중흥S클래스                  613.400000
성판악 → 수악교                              562.773234
수악교 → 성판악                              551.072100
시티주유소 → 제주한라대학교(종점)                    423.162996
탐라도서관 → 으뜸마을                           414.000000
제주도청 신제주로터리 → 제주국제공항(구제주방면)            360.272393
고성리제주은행 → 성산환승정류장(고성리 회전교차로)           327.722892
성판악 → 교래입구                             314.931116
고성리 구 성산농협 → 성산환승정류장(고성리 한국물류)         302.235294
교래입구 → 성판악                             301.922481
제주국제공항(구제주방면) → 월성 마을                  291.454068
고성리 성산농협 → 성산환승정류장(고성리 한국물류)      

In [36]:
up = (33.506280, 126.490312) # 제주공항
center = (33.379724, 126.545315) # 한라산

In [37]:
station_location_df = df.copy()

In [38]:
station_location_df

,id,date,route_id,vh_id,route_nm,now_latitude,now_longitude,now_station,now_arrive_time,distance,next_station,next_latitude,next_longitude,next_arrive_time,now_hour,dayofweek,station_segment
0,0,2019-10-15,405136001,7997025,360-1,33.456267,126.551750,제주대학교입구,06시,266.0,제대마을,33.457724,126.554014,24,6,1,제주대학교입구 → 제대마을
1,1,2019-10-15,405136001,7997025,360-1,33.457724,126.554014,제대마을,06시,333.0,제대아파트,33.458783,126.557353,36,6,1,제대마을 → 제대아파트
2,2,2019-10-15,405136001,7997025,360-1,33.458783,126.557353,제대아파트,06시,415.0,제주대학교,33.459893,126.561624,40,6,1,제대아파트 → 제주대학교
3,3,2019-10-15,405136001,7997025,360-1,33.479705,126.543811,남국원(아라방면),06시,578.0,제주여자중고등학교(아라방면),33.484860,126.542928,42,6,1,남국원(아라방면) → 제주여자중고등학교(아라방면)
4,4,2019-10-15,405136001,7997025,360-1,33.485662,126.494923,도호동,07시,374.0,은남동,33.485822,126.490897,64,7,1,도호동 → 은남동
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210452,210452,2019-10-28,405328102,7983486,281-2,33.255783,126.577450,비석거리,21시,528.0,삼아아파트,33.251896,126.574417,96,21,0,비석거리 → 삼아아파트
210453,210453,2019-10-28,405328102,7983486,281-2,33.248595,126.568527,동문로터리,21시,280.0,매일올레시장 7번입구,33.249753,126.565959,50,21,0,동문로터리 → 매일올레시장 7번입구
210454,210454,2019-10-28,405328102,7983486,281-2,33.251891,126.560303,서귀포시 구 버스터미널,21시,114.0,아랑조을거리 입구,33.251084,126.559551,16,21,0,서귀포시 구 버스터미널 → 아랑조을거리 입구
210455,210455,2019-10-28,405328102,7983486,281-2,33.251084,126.559551,아랑조을거리 입구,21시,223.0,평생학습관,33.249504,126.558068,38,21,0,아랑조을거리 입구 → 평생학습관


for _, row in station_location_df.iterrows():
    #print(f'idx = {idx}')
    print(f'row = {row}')
    print('=' * 30)

In [39]:
heat_data = station_location_df[['now_latitude', 'now_longitude']].values.tolist()

jeju_map = folium.Map(
            location = [33.38, 126.55],
            zoom_start = 10
            )

HeatMap(heat_data).add_to(jeju_map)
# folium.Marker(
#     location = [up[0], up[1]],
#     popup = '공항!!', tooltip = "제주공항",
#     icon = folium.Icon(color='red')    
# ).add_to(jeju_map)

for _, row in station_location_df.iterrows():
    folium.CircleMarker(
        location = [row['now_latitude'], row['now_longitude']],
        radius = 2,
        fill = True,
        popup = row['now_station']
    ).add_to(jeju_map)

jeju_map

In [41]:
analysis_df = df.copy()

In [42]:
long_time_df = analysis_df[  analysis_df['next_arrive_time'] >= 340  ]

In [43]:
top_50 = long_time_df.sort_values(by = 'next_arrive_time', ascending = False).head(50)

jeju_map = folium.Map(location = [33.38, 126.55])

for idx, row in top_50.iterrows():
    folium.Marker(location = [row['now_latitude'], row['now_longitude']], tooltip= row['now_station']
    ).add_to(jeju_map)
jeju_map